backends/ebpf/psa/externs/ebpfPsaTableImplementation.cpp

```diff
@@ -636,4 +654,137 @@ void EBPFActionSelectorPSA::verifyTableEmptyGroupAction(const EBPFTablePSA* inst
     }
 }
 
+void EBPFActionSelectorPSA::emitCacheTypes(CodeBuilder* builder) {
+    if (!tableCacheEnabled) return;
+
+    CodeGenInspector commentGen(program->refMap, program->typeMap);
+    commentGen.setBuilder(builder);
+
+    // construct cache key type - we need group reference and selector keys
+    builder->emitIndent();
+    builder->appendFormat("struct %s ", cacheKeyTypeName.c_str());
+    builder->blockStart();
+
+    builder->emitIndent();
+    builder->appendFormat("u32 group_ref");
+    builder->endOfStatement(true);
+
+    unsigned int fieldNumber = 0;
+    for (auto s : selectors) {
+        auto type = program->typeMap->getType(s->expression);
+        auto ebpfType = EBPFTypeFactory::instance->create(type);
+        cstring fieldName = Util::printf_format("field%u", fieldNumber++);
+
+        builder->emitIndent();
+        ebpfType->declare(builder, fieldName, false);
+        builder->endOfStatement(false);
+        builder->append(" /* ");
+        s->expression->apply(commentGen);
+        builder->append(" */");
+        builder->newline();
+    }
+
+    builder->blockEnd(false);
+    builder->append(" __attribute__((aligned(4)))");
+    builder->endOfStatement(true);
+
+    // value can be used from Action Selector because there is no need to cache hit variable
+}
+
+void EBPFActionSelectorPSA::emitCacheVariables(CodeBuilder* builder) {
+    if (!tableCacheEnabled) return;
+
+    cacheKeyVar = program->refMap->newName("key_cache");
+    cacheDoUpdateVar = program->refMap->newName("do_update_cache");
+
+    builder->emitIndent();
+    builder->appendFormat("struct %s %s = {0}", cacheKeyTypeName.c_str(), cacheKeyVar.c_str());
+    builder->endOfStatement(true);
+
+    builder->emitIndent();
+    builder->appendFormat("u8 %s = 0", cacheDoUpdateVar.c_str());
+    builder->endOfStatement(true);
+}
+
+void EBPFActionSelectorPSA::emitCacheLookup(CodeBuilder* builder, cstring key, cstring value) {
+    if (!tableCacheEnabled) return;
+
+    builder->emitIndent();
+    builder->appendFormat("%s.group_ref = %s->%s", cacheKeyVar.c_str(), key.c_str(),
+                          referenceName.c_str());
+    builder->endOfStatement(true);
+
+    unsigned int fieldNumber = 0;
+    for (auto s : selectors) {
+        auto type = program->typeMap->getType(s->expression);
+        auto ebpfType = EBPFTypeFactory::instance->create(type);
+        cstring fieldName = Util::printf_format("field%u", fieldNumber++);
+
+        bool memcpy = false;
+        auto scalar = ebpfType->to<EBPFScalarType>();
+        unsigned width = 0;
+        if (scalar != nullptr) {
+            width = scalar->implementationWidthInBits();
+            memcpy = !EBPFScalarType::generatesScalar(width);
+        }
+
+        builder->emitIndent();
+        if (memcpy) {
+            builder->appendFormat("memcpy(&%s.%s, &", cacheKeyVar.c_str(), fieldName.c_str());
```

Comment:
so ebpf supports memcpy... or perhaps this is just an intrinsic that the compiler knows about

Reply:
Probably it maps to an intrinsic that the compiler knows about. I forgot to change it to `__builtin_memcpy`.